airbyte-workload-launcher/src/main/kotlin/ClaimedProcessor.kt

```diff
@@ -94,26 +96,33 @@ class ClaimedProcessor(
   private fun getWorkloadList(workloadListRequest: WorkloadListRequest): WorkloadListResponse {
     while (true) {
       try {
-        // TODO: consider tuning the retry policy here, since we currently get the default 2 retries.
         return Failsafe.with(
           RetryPolicy.builder<Any>()
-            .withBackoff(Duration.ofSeconds(20), Duration.ofDays(365))
+            .withBackoff(Duration.ofSeconds(20), Duration.ofMinutes(10))
```

Comment:
Hi @gosusnp, thanks for taking the time to review this.

From my perspective, since `ClaimedProcessor.getWorkloadList()` is not optional for the pod to function, but rather _sine qua non_,  I believe it is fair to fail after some threshold, similar to how pipelines fail after X retries. This brings more transparency on the real status of the pod, instead of just logging that it didn't connect for 365 days. What is the ideal threshold? I am not sure. I set 10 minutes because this is "usually" more than enough time for a node to recover and be ready to accept pods in a k8s cluster, but maybe it should/could be longer, or maybe it should be governed by a ENV VAR (ideal setup in my opinion - to give the developer the choice with a standard value of 365 days if required).

On the other hand, I do understand that it seems that this piece of code is relevant for other aspects of the project for reasons that are not 100% clear to me.

If we implement the retry mechanism for `SocketException` alone (without the 10 minutes fail throw or an ENV VAR for this), it is, in my view, already a win, with no apparent downsides.

Reply:
Hi @gosusnp, thanks for taking the time to review this.

True, I missed the maximum retry policy, the way it was proposed just set the maximum back off to 10 minutes, but still retries forever (I would need to add a .withMaxDuration(Duration.ofXXX(X))). I also got your view that you think this is not necessary, and it should be retried forever.

Yet, from my perspective, since ClaimedProcessor.getWorkloadList() is not optional for the pod to function, but rather sine qua non, I believe it is fair to fail after some threshold, similar to how pipelines fail after X retries. This brings more transparency on the real status of the pod, instead of just logging that it didn't connect for 20s backoffs until 365 days, and every 365 days after. What is the ideal threshold? I am not sure.

I do understand that it seems that this forever retry is relevant for other aspects of the project for reasons that are not 100% clear to me (although I believe that nothing should retry forever - after a given limit it should stop and throw an error for transparency and resource optimization - but this is just my opinion).

Perhaps the ideal would be to create ENV VARs with backoff, maximum backoff, and maximum retry time limit, so that developers can have the option of what is their ideal threshold (exposing in the Helm chart).

Furthermore, if we implement the retry mechanism for SocketException alone (without the 10 minutes fail throw or an ENV VAR for this), it is, in my view, already a win, with no apparent downsides.

Reply:
I view the dependency of the `workload-launcher` on the `workfload-api` the same as an api-server on a database. In the api-server / database situation, you don't want your liveness to depend on the db to avoid churning through all the pods if the db was to become unavailable. I believe this is true for the launcher and the workload-api as well.

Now, I am interested in trying to understand why you are running into this situation more than we do. What I suspect is that we might have a couple of issues, I wonder if one of them is that the launcher starts before the workload-api.
Would you mind sharing some context into what operations you are doing when running into this?

Reply:
The only operation I do when I run into this is when I do a new release.

As you've mentioned, most likely the workload-api is not yet ready by the time workload-launcher do the request, and without a retry it cannot recover, forcing me to manually kill the pod.

Reply:
If the pod is not killed, the pipelines won't run, and will retry as long as they can.

Killing the pod instantly recovers Airbyte once it is back.